src/memray/commands/attach.py

```diff
@@ -0,0 +1,377 @@
+        f"-ex=set $port={port}",
+        f"-x={GDB_SCRIPT}",
+    ]
+
+    # When adding new breakpoints, also update _attach.gdb
+    lldb_breakpoint_args = [
+        "breakpoint",
+        "set",
+        "-b",
+        "malloc",
+        "-b",
+        "calloc",
+        "-b",
+        "realloc",
+        "-b",
+        "free",
+        "-b",
+        "PyMem_Malloc",
+        "-b",
+        "PyMem_Calloc",
+        "-b",
+        "PyMem_Realloc",
+        "-b",
+        "PyMem_Free",
+        "-C",
+        "'breakpoint disable'",
+        "-C",
+        f"'expr auto $rtld_default = (void*){RTLD_DEFAULT}'",
+        "-C",
+        f"'expr auto $rtld_now = {RTLD_NOW}'",
+        "-C",
+        "'expr auto $dlsym = (void* (*)(void*, const char*))&dlsym'",
+        "-C",
+        """'expr auto $dlopen = $dlsym($rtld_default, "dlopen")'""",
+        "-C",
+        """'expr auto $dlerror = $dlsym($rtld_default, "dlerror")'""",
+        "-C",
+        "'expr auto $dll = ((void*(*)(const char*, int))$dlopen)($libpath, $rtld_now)'",
+        "-C",
+        "'p ((char*(*)(void))$dlerror)()'",
+        "-C",
+        """'expr auto $spawn = $dlsym($dll, "memray_spawn_client")'""",
+        "-C",
+        "'p ((const char*(*)(int))$spawn)($port)'",
+    ]
+
+    lldb_cmd = [
+        "lldb",
+        "--batch",
+        "-p",
+        str(pid),
+        "--no-lldbinit",
+        "--source-quietly",
+        "-o",
+        "p ((void*(*)(size_t))PyMem_Malloc)",
+        "-o",
+        "p ((void*(*)(size_t, size_t))PyMem_Calloc)",
+        "-o",
+        "p ((void*(*)(void *, size_t))PyMem_Realloc)",
+        "-o",
+        "p ((void(*)(void*))PyMem_Free)",
+        "-o",
+        f'expr char $libpath[]="{injecter}"',
+        "-o",
+        f"expr int $port={port}",
+        "-o",
+        " ".join(lldb_breakpoint_args),
+        "-o",
+        "continue",
+    ]
+
+    cmd = gdb_cmd if debugger == "gdb" else lldb_cmd
+    # print(shlex.join(cmd))
+    p = subprocess.Popen(
+        cmd,
+        text=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+    )
+    output = p.communicate()[0]
+    # print(f"debugger return code: {p.returncode}")
+    # print(f"debugger output:\n{output}")
+    return (
+        p.returncode == 0
+        and "error:" not in output
+        and "unrecognized option" not in output
+        and "Operation not permitted" not in output
+        and "attach failed" not in output
+    )
+
+
+def _gdb_available() -> bool:
+    # We require at least gdb 8.0 or our CI can fail due to
+    # https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=898048
+    try:
+        version_text = subprocess.check_output(["gdb", "--version"], text=True)
+    except FileNotFoundError:
+        return False
+    except subprocess.CalledProcessError:
+        return False
+
+    lines = version_text.splitlines()
+    if not lines:
+        return False
+
+    version_words = lines[0].strip().split()
```

Comment:
I have no idea why I didn't think of using a regex for this. It would definitely be less work.

Another option is using this instead:
```
$ gdb -batch -iex 'p $_gdb_major'
$1 = 12
```

_But_, that was only added in GDB 9, and we ought to be compatible with GDB 8 as well. The most resilient version of this check might be checking whether `$_gdb_major >= 9` OR `"8." in check_output("gdb --help")`

Honestly... I added this because CI is running on platforms where gdb 7 happens to be broken, but that's a gdb bug, not a memray issue at all. And there are machines where memray works just fine with gdb 7 - as long as they're not running on a CPU that gdb 7 has broken support for.

I'm thinking the best option is removing this check from `memray attach`, and instead trying to handle it as a special case in the test suite. What do you think?

Reply:
OK, I've talked myself into doing this differently. I'm going to make `memray attach` itself accept any `gdb` version, and I'll make the test suite xfail if the debugger output contains "Couldn't write extended state status" (since we see that when we're being bitten by https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=898048 in CI).